src/Workspaces/Remote/Core/RemoteEndPoint.cs

```diff
@@ -196,16 +196,22 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object?> ar
                 // if invoke throws an exception, make sure we raise cancellation.
                 RaiseCancellationIfInvokeFailed(task, linkedCancellationSource, cancellationToken);
 
-                // wait for asset source to respond
-                await pipe.WaitForConnectionAsync(linkedCancellationSource.Token).ConfigureAwait(false);
+                var task2 = Task.Run(async () =>
+                {
+                    // wait for asset source to respond
+                    await pipe.WaitForConnectionAsync(linkedCancellationSource.Token).ConfigureAwait(false);
 
-                // run user task with direct stream
-                var result = await dataReader(stream, linkedCancellationSource.Token).ConfigureAwait(false);
+                    // run user task with direct stream
+                    return await dataReader(stream, linkedCancellationSource.Token).ConfigureAwait(false);
+                }, linkedCancellationSource.Token);
 
-                // wait task to finish
-                await task.ConfigureAwait(false);
+                // Wait for all tasks to finish. If we return while one of the tasks is still in flight, the task would
+                // operate as a fire-and-forget operation where the caller might release state objects still in use by
+                // the task.
+                await Task.WhenAll(task, task2).ConfigureAwait(false);
```

Comment:
wasn't this change already done?  this looks super familiar.

Reply:
This is a back-port of #49164 to release/dev16.8

Reply:
ah ok :)